SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/LayeredLBStrategy.cpp

```diff
@@ -0,0 +1,278 @@
+			UAbstractLBStrategy* DefaultLBStrategy = NewObject<UAbstractLBStrategy>(this, WorldSettings->LoadBalanceStrategy);
+			AddStrategyForLayer(SpatialConstants::DefaultLayer, DefaultLBStrategy);
+		}
+	}
+}
+
+void ULayeredLBStrategy::SetLocalVirtualWorkerId(VirtualWorkerId InLocalVirtualWorkerId)
+{
+	LocalVirtualWorkerId = InLocalVirtualWorkerId;
+	for (const auto& Elem : LayerNameToLBStrategy)
+	{
+		Elem.Value->SetLocalVirtualWorkerId(InLocalVirtualWorkerId);
+	}
+}
+
+bool ULayeredLBStrategy::ShouldHaveAuthority(const AActor& Actor) const
+{
+	if (!IsReady())
+	{
+		UE_LOG(LogLayeredLBStrategy, Warning, TEXT("LayeredLBStrategy not ready to relinquish authority for Actor %s."), *AActor::GetDebugName(&Actor));
+		return false;
+	}
+
+	const FName& LayerName = GetLayerNameForActor(Actor);
+	if (!LayerNameToLBStrategy.Contains(LayerName))
+	{
+		UE_LOG(LogLayeredLBStrategy, Warning, TEXT("LayeredLBStrategy doesn't have a LBStrategy for Actor %s which is in Layer %s."), *AActor::GetDebugName(&Actor), *LayerName.ToString());
+		return false;
+	}
+
+	// If this worker is not responsible for the Actor's layer, just return false.
+	if (VirtualWorkerIdToLayerName.Contains(LocalVirtualWorkerId) && VirtualWorkerIdToLayerName[LocalVirtualWorkerId] != LayerName)
+	{
+		return false;
+	}
+
+	return LayerNameToLBStrategy[LayerName]->ShouldHaveAuthority(Actor);
+}
+
+VirtualWorkerId ULayeredLBStrategy::WhoShouldHaveAuthority(const AActor& Actor) const
+{
+	if (!IsReady())
+	{
+		UE_LOG(LogLayeredLBStrategy, Warning, TEXT("LayeredLBStrategy not ready to decide on authority for Actor %s."), *AActor::GetDebugName(&Actor));
+		return SpatialConstants::INVALID_VIRTUAL_WORKER_ID;
+	}
+
+	const FName& LayerName = GetLayerNameForActor(Actor);
+	if (!LayerNameToLBStrategy.Contains(LayerName))
+	{
+		UE_LOG(LogLayeredLBStrategy, Warning, TEXT("LayeredLBStrategy doesn't have a LBStrategy for Actor %s which is in Layer %s."), *AActor::GetDebugName(&Actor), *LayerName.ToString());
+		return SpatialConstants::INVALID_VIRTUAL_WORKER_ID;
+	}
+
+	const VirtualWorkerId ReturnedWorkerId = LayerNameToLBStrategy[LayerName]->WhoShouldHaveAuthority(Actor);
+
+	UE_LOG(LogLayeredLBStrategy, Log, TEXT("LayeredLBStrategy returning virtual worker id %d for Actor %s."), ReturnedWorkerId, *AActor::GetDebugName(&Actor));
+	return ReturnedWorkerId;
+}
+
+SpatialGDK::QueryConstraint ULayeredLBStrategy::GetWorkerInterestQueryConstraint() const
+{
+	check(IsReady());
+	if (!VirtualWorkerIdToLayerName.Contains(LocalVirtualWorkerId))
+	{
+		UE_LOG(LogLayeredLBStrategy, Warning, TEXT("LayeredLBStrategy doesn't have a LBStrategy for worker %d."), LocalVirtualWorkerId);
+		SpatialGDK::QueryConstraint Constraint;
+		Constraint.ComponentConstraint = 0;
+		return Constraint;
+	}
+	else
+	{
+		const FName& LayerName = VirtualWorkerIdToLayerName[LocalVirtualWorkerId];
+		check(LayerNameToLBStrategy.Contains(LayerName));
+		return LayerNameToLBStrategy[LayerName]->GetWorkerInterestQueryConstraint();
+	}
```

Comment:
NIT: we return at the end of the error-case if statement above so this can be de-indented (un-indented? antidented?)

Reply:
Personally I find this easier to read when it's two equal cases?

Reply:
Valentyn prefers it this way too XD.
This is pure preference I think.

Reply:
I think the reasoning behind exiting in the error case and avoiding the `else {...}` indenting is that if you've got multiple error possibilities in the same function you end up with X levels of indentation. 

Totally accept it's preference, though apologies in advance if I forget about this conversation in six months time and refactor it 😄